src/vmm/src/lib.rs

```diff
@@ -334,6 +334,20 @@ impl Vmm {
         self.mmio_device_manager.kick_devices();
         self.broadcast_vcpu_event(VcpuEvent::Resume, VcpuResponse::Resumed)
             .map_err(|_| Error::VcpuResume)?;
+        // When restoring from a previously saved state, there is no serial
+        // driver initialization, therefore the RDA (Received Data Available)
+        // interrupt is not enabled. Because of that, the driver won't get
+        // notified of any bytes that we send to the guest. The clean solution
+        // would be to save the whole serial device state when we do the vm
+        // serialization. For now we set that bit manually
+        #[cfg(target_arch = "x86_64")]
+        self.pio_device_manager
+            .stdio_serial
+            .lock()
+            .expect("Poisoned lock")
+            .serial
+            .write(1, 0b0000_0001)
+            .map_err(|_| Error::Serial(std::io::Error::last_os_error()))?;
```

Comment:
Why not set it for `x86_64` in `MMIODeviceManager::restore()` just like you do for `aarch64`? It's weird to separate them like this..

Reply:
I agree. It's also that resume is also called when resuming a paused VM (not neccessarily a snapshotted one).

Reply:
Maybe in `builder::build_from_snapshot`? After building the PIO Serial on x86 and MMIO Serial on Arm you can call an `emulate_serial_driver_init()` specialized fn. wdyt?

Reply:
how is this different from how we kick all devices after resume?

Reply:
> how is this different from how we kick all devices after resume?

well.. that was a _necessary_ (ugly) hack coming from the fact that `EventManager` events are not saved to snapshot.

I am hoping there is a nicer way to solve this for the Serial :pray:

Reply:
I guess we can put it in `build_from_snapshot` does it seem better?